### ISSUE TITLE
first submodule import always binds the submodule name, even if it was previously imported in another file

### ISSUE THREAD

**Issue by jelle-openai** [astral-sh/ty#4553]
### Summary

This package runs successfully in Python and passes ty 0.0.81, but ty 0.0.82 reports `call-non-callable`.

Create these three files in an empty directory:

`capybara/__init__.py`
```python
from capybara.gundi import hutia
from capybara.hutia import vole
hutia()
```

`capybara/gundi.py`
```python
import capybara.hutia

def hutia():
    pass
```

`capybara/hutia.py`
```python
vole = 0
```

From that directory, run `python -B -c 'import capybara'` and `ty check`.

Expected: both commands succeed. The `hutia` submodule is loaded while importing `gundi`, before `__init__` binds `hutia` to the exported function. The subsequent import from that already-loaded submodule does not replace the function binding at runtime.

Actual: Python succeeds, and ty 0.0.81 reports `All checks passed!`. ty 0.0.82 reports `Object of type <module 'capybara.hutia'> is not callable` at `hutia()`.

This was reduced from Streamlit 1.56.0's `navigation` re-export. The final reproducer needs no third-party dependencies. Removing either import in `__init__`, or moving the helper function into the package initializer, removes the version difference. Removing the preload import from `gundi` loses the runtime-valid behavior, so that import is essential.

**Comment by carljm** [comment]
It's not clear what the best option here is. Runtime-accurate modeling of this import side effect is not really feasible, because it depends on global import order. This code works iff `capybara.hutia` has been imported before `from capybara.hutia import vole` runs. If we remove `import capybara.hutia` from `capybara/gundi.py`, then this code fails at runtime.

Mypy models implicit submodule imports in such a way that they never override explicit bindings of the same name. Perhaps that is the best option. It means that we would fail to catch the error if `import capybara.hutia` were removed, but that's probably a better tradeoff.

**Comment by carljm** [comment]
The question is basically whether it's better for a type checker to conservatively alert you to the fact that your code implicitly depends on global import order in a fragile, non-local way, or it's better for a type checker to try to avoid false positives and assume that your code probably works or you'd have noticed runtime failures.

**Comment by carljm** [comment]
It is also in principle possible that we could scan the full transitive closure of import dependencies, checking to see whether we can confirm that a prior import of `capybara.hutia` must have occurred. This would allow us to avoid false positives in any case where the code definitely works (no matter how far away the import is that makes that true) while still warning in case the working code is actually non-deterministic based on external import order changes. But this could be quite an expensive walk, so I'm not sure this approach is practically feasible.

**Comment by AlexWaygood** [comment]
Does https://github.com/astral-sh/ruff/pull/21583, or any of the other old submodule-import experiments I did with @Gankra that I never ended up getting over the line, help with this at all?

**Comment by carljm** [comment]
> Does [astral-sh/ruff#21583](https://github.com/astral-sh/ruff/pull/21583), or any of the other old submodule-import experiments I did with [@Gankra](https://github.com/Gankra) that I never ended up getting over the line, help with this at all?

I don't know, I suggest asking Codex :)

**Comment by AlexWaygood** [comment]
per codex:

---

Neither https://github.com/astral-sh/ruff/pull/21583 nor its follow-up https://github.com/astral-sh/ruff/pull/21653 fixes this reproducer. I verified by rebasing #21583 onto current `main` (`5bb508f594`), resolving conflicts from intervening refactors, then rebasing #21653 onto that. Freshly built binaries from both branches still report `call-non-callable` on the exact three-file example.

The distinction is that these PRs change **module attribute lookup**, whereas this example fails because ty creates a **local binding inside `__init__.py`**. Both rebased branches produce:

```python
from typing import reveal_type

from capybara.gundi import hutia
reveal_type(hutia)  # def hutia() -> Unknown

from capybara.hutia import vole
reveal_type(hutia)  # <module 'capybara.hutia'>
```

The second import synthesizes a local `hutia` module binding, overwriting the function. That mechanism does not recognize the earlier transitive import in `gundi.py`, and neither PR changes it. Applying the “implicit imports never override explicit bindings” approach here would require changing that local-binding mechanism too.

I also checked separate control cases to confirm that the PRs’ intended behavior changes survived the rebase: #21583 makes submodules available through `from` imports, and #21653 lets explicit package attributes take precedence over submodules. Those changes work, but they do not affect this particular case.

**Comment by jelle-openai** [comment]
I think this caused exactly one hit in my codebase and agree the pattern is pretty questionable, so I'd be OK putting this on the back burner.

**Comment by ppentchev** [comment]
FWIW, this also happens with the [`cappa` command-line parsing library](https://pypi.org/project/cappa) - its `__init__.py` module imports at least `command` and `parse` as functions from its submodules, and there are `cappa.command` and `cappa.parse` "hidden" submodules. I could try to get in touch with the author, he is open to reasonable discussion, or, of course, I could override this in my projects that use `cappa`, but just thought I'd point it out.

See [this media-type-version change](https://git.sr.ht/~ppentchev/media-type-version/commit/d3a443a01e7a7a63b7e703d5388eb42f36e6db10) for an example of the needed overrides.
